cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewRunner.java

```diff
@@ -0,0 +1,213 @@
+  private ProgramId programId;
+
+  @Inject
+  DefaultPreviewRunner(DatasetService datasetService, LogAppenderInitializer logAppenderInitializer,
+                       ApplicationLifecycleService applicationLifecycleService,
+                       SystemArtifactLoader systemArtifactLoader, ProgramRuntimeService programRuntimeService,
+                       ProgramLifecycleService programLifecycleService, DefaultNamespaceEnsurer namespaceEnsurer,
+                       PreviewStore previewStore) {
+    this.datasetService = datasetService;
+    this.logAppenderInitializer = logAppenderInitializer;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.systemArtifactLoader = systemArtifactLoader;
+    this.programRuntimeService = programRuntimeService;
+    this.programLifecycleService = programLifecycleService;
+    this.namespaceEnsurer = namespaceEnsurer;
+    this.previewStore = previewStore;
+    this.status = null;
+  }
+
+  @Override
+  public void startPreview(PreviewRequest<?> previewRequest) throws Exception {
+    programId = previewRequest.getProgram();
+    AppRequest<?> request = previewRequest.getAppRequest();
+    ArtifactSummary artifactSummary = request.getArtifact();
+    ApplicationId preview = programId.getParent();
+    NamespaceId artifactNamespace = ArtifactScope.SYSTEM.equals((artifactSummary.getScope())) ? NamespaceId.SYSTEM
+      : preview.getParent();
+
+    ArtifactId artifactId = new ArtifactId(artifactNamespace.getNamespace(), artifactSummary.getName(),
+                                           artifactSummary.getVersion());
+
+    String config = request.getConfig() == null ? null : GSON.toJson(request.getConfig());
+
+    try {
+      applicationLifecycleService.deployApp(preview.getParent(), preview.getApplication(), preview.getVersion(),
+                                            artifactId.toId(), config, NOOP_PROGRAM_TERMINATOR);
+    } catch (Exception e) {
+      this.status = new PreviewStatus(PreviewStatus.Status.DEPLOY_FAILED, new BasicThrowable(e));
+      throw e;
+    }
+
+    ProgramController controller = programLifecycleService.start(programId, new HashMap<String, String>(), false);
+
+    controller.addListener(new AbstractListener() {
+      @Override
+      public void init(ProgramController.State currentState, @Nullable Throwable cause) {
+        status = new PreviewStatus(PreviewStatus.Status.RUNNING, null);
+      }
+
+      @Override
+      public void completed() {
+        setStatus(new PreviewStatus(PreviewStatus.Status.COMPLETED, null));
+      }
+
+      @Override
+      public void killed() {
+        setStatus(new PreviewStatus(PreviewStatus.Status.KILLED, null));
+      }
+
+      @Override
+      public void error(Throwable cause) {
+        setStatus(new PreviewStatus(PreviewStatus.Status.RUN_FAILED, new BasicThrowable(cause)));
+      }
+    }, Threads.SAME_THREAD_EXECUTOR);
+  }
+
+  private synchronized void setStatus(PreviewStatus status) {
```

Comment:
Usually private method doesn't need to be synchronized (and it is not needed in this case).